tests/ndc/ndc.go

```diff
@@ -0,0 +1,2468 @@
+		}, nil).AnyTimes()
+	mockOtherClient.EXPECT().StreamWorkflowReplicationMessages(gomock.Any()).Return(mockStreamClient, nil).AnyTimes()
+	s.mockAdminClient = map[string]adminservice.AdminServiceClient{
+		"cluster-b": mockStandbyClient,
+		"cluster-c": mockOtherClient,
+	}
+	clusterConfigs[0].MockAdminClient = s.mockAdminClient
+
+	cluster, err := s.testClusterFactory.NewCluster(s.T(), clusterConfigs[0], log.With(s.logger, tag.ClusterName(clusterName[0])))
+	s.Require().NoError(err)
+	s.cluster = cluster
+
+	s.registerNamespace()
+
+	s.version = clusterConfigs[1].ClusterMetadata.ClusterInformation[clusterConfigs[1].ClusterMetadata.CurrentClusterName].InitialFailoverVersion
+	s.versionIncrement = clusterConfigs[0].ClusterMetadata.FailoverVersionIncrement
+	s.generator = test.InitializeHistoryEventGenerator(s.namespace, s.namespaceID, s.version)
+}
+
+func (s *NDCFunctionalTestSuite) GetReplicationMessagesMock(
+	ctx context.Context,
+	request *adminservice.GetReplicationMessagesRequest,
+	opts ...grpc.CallOption,
+) (*adminservice.GetReplicationMessagesResponse, error) {
+	select {
+	case task := <-s.standByReplicationTasksChan:
+		taskID := atomic.AddInt64(&s.standByTaskID, 1)
+		task.SourceTaskId = taskID
+		tasks := []*replicationspb.ReplicationTask{task}
+		for len(s.standByReplicationTasksChan) > 0 {
+			task = <-s.standByReplicationTasksChan
+			taskID := atomic.AddInt64(&s.standByTaskID, 1)
+			task.SourceTaskId = taskID
+			tasks = append(tasks, task)
+		}
+
+		replicationMessage := &replicationspb.ReplicationMessages{
+			ReplicationTasks:       tasks,
+			LastRetrievedMessageId: tasks[len(tasks)-1].SourceTaskId,
+			HasMore:                true,
+		}
+
+		return &adminservice.GetReplicationMessagesResponse{
+			ShardMessages: map[int32]*replicationspb.ReplicationMessages{1: replicationMessage},
+		}, nil
+	default:
+		return &adminservice.GetReplicationMessagesResponse{
+			ShardMessages: make(map[int32]*replicationspb.ReplicationMessages),
+		}, nil
+	}
+}
+
+func (s *NDCFunctionalTestSuite) SetupTest() {
+	// Have to define our overridden assertions in the test setup. If we did it earlier, s.T() will return nil
+	s.Assertions = require.New(s.T())
+	s.ProtoAssertions = protorequire.New(s.T())
+	s.generator = test.InitializeHistoryEventGenerator(s.namespace, s.namespaceID, s.version)
+}
+
+func (s *NDCFunctionalTestSuite) TearDownSuite() {
+	if s.generator != nil {
+		s.generator.Reset()
+	}
+	s.controller.Finish()
+	s.NoError(s.cluster.TearDownCluster())
+}
+
+func (s *NDCFunctionalTestSuite) TestSingleBranch() {
+
+	s.setupRemoteFrontendClients()
+	workflowID := "ndc-single-branch-test" + uuid.New()
+
+	workflowType := "event-generator-workflow-type"
+	taskqueue := "event-generator-taskQueue"
+
+	// cluster has initial version 1
+	historyClient := s.cluster.GetHistoryClient()
+
+	versions := []int64{3, 13, 2, 202, 302, 402, 602, 502, 802, 1002, 902, 702, 1102}
+	for _, version := range versions {
```

Comment:
It looks like you are testing each version separately. Is that correct? What is the reason not to use t.Run for each sub-test?